x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/Predicates.java

```diff
@@ -33,15 +36,38 @@
             list.addAll(splitOr(or.right()));
             return list;
         }
-        return Collections.singletonList(exp);
+        return singletonList(exp);
     }
 
     public static Expression combineOr(List<Expression> exps) {
-        return exps.stream().reduce((l, r) -> new Or(l.location(), l, r)).orElse(null);
+        return combine(exps, (l, r) -> new Or(l.location(), l, r));
     }
 
     public static Expression combineAnd(List<Expression> exps) {
-        return exps.stream().reduce((l, r) -> new And(l.location(), l, r)).orElse(null);
+        return combine(exps, (l, r) -> new And(l.location(), l, r));
+    }
+
+    /**
+     * Build a binary 'pyramid' - while a bit longer this method creates a balanced tree as oppose to a plain
+     * recursive approach that creates an unbalanced one (either to the left or right).
+     */
+    private static Expression combine(List<Expression> exps, BiFunction<Expression, Expression, Expression> combiner) {
+        if (exps.isEmpty()) {
+            return null;
+        }
+
+        List<Expression> result = new ArrayList<>(exps);
```

Comment:
I talked with @costin earlier about this - he wants to keep the order the same and my proposal doesn't.

What about this?
```
while (result.size() > 1) {
  ListIterator<Expression> itr = result.iterator();
  while (itr.hasNext()) {
    itr.add(combiner.apply(itr.remove(), itr.remove()));
  }
}
```


Your version works but `for (int i = 0; i < result.size() - 1; i++) {` make me think it'll be a normal loop and then you remove and add and I'm confused. Using the `ListIterator` forces the reader to think.

Reply:
If you'd like we can follow this up in a different PR.
The approach above doesn't work since `remove` can only be called _once_ per _next_/_previous_.

I've added a comment to the loop explaining that it updates the list (and thus why it uses the temporary variable).
Considering the loops are fairly contained I think that conveys the message without the gymnastics of iterator.

P.S. The list is created just above as a copy for this specific reason.